src/OrleansServiceFabricUtils/Utilities/WaitHandleExtensions.cs

```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Orleans.ServiceFabric.Utilities
+{
+    internal static class WaitHandleExtensions
+    {
+        internal static Task<bool> WaitAsync(this WaitHandle handle, TimeSpan timeout)
+        {
+            // If the handle is already set, return synchronously.
+            if (handle.WaitOne(0))
+                return Task.FromResult(true);
+
+            // If the timeout is zero, return synchronously.
+            if (timeout == TimeSpan.Zero)
+                return Task.FromResult(false);
+            
+            // Register a delegate on the thread pool to execute when the handle is set.
+            var tcs = new TaskCompletionSource<bool>();
+            ThreadPool.RegisterWaitForSingleObject(
```

Comment:
What's the purpose of using this extension in conjunction with a WaitHandle, instead of just relying on a TCS? This way you skip the wait handle, the extra blocked thread, etc